userspace/otpilot/src/flash.rs

```diff
@@ -0,0 +1,174 @@
+// Copyright 2021 lowRISC contributors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+use core::cell::Cell;
+
+use libtock::result::TockResult;
+use libtock::syscalls;
+use libtock::syscalls::raw::yieldk;
+
+pub const MAX_BUFFER_LENGTH: usize = 128;
+
+pub trait Flash {
+    // Read from flash.
+    // offset: Location relative to flash start to read from. Must be word (4 bytes) aligned.
+    // buffer: Buffer for read data.
+    //         It must be: buffer.len() <= len <= MAX_BUFFER_LENGTH
+    // read_length: Number of bytes to read. Must be > 0 and a multiple of 4.
+    fn read(&self, offset: usize, buffer: &mut[u8], len: usize) -> TockResult<()>;
+
+    // Write to flash.
+    // offset: Location relative to flash start to write to. Must be word (4 bytes) aligned.
+    // buffer: Buffer with data to write.
+    //         It must be: buffer.len() <= len <= MAX_BUFFER_LENGTH
+    // read_length: Number of bytes to write. Must be > 0 and a multiple of 4.
+    fn write(&self, offset: usize, buffer: &mut[u8], len: usize) -> TockResult<()>;
+
+    // Erase page in flash.
+    // page: Number of page to erase.
+    fn erase(&self, page: usize) -> TockResult<()>;
+
+    // Returns true if the last operation is done.
+    fn is_operation_done(&self) -> bool;
+
+    // Wait (yieldk) until the operation is done.
+    fn wait_operation_done(&self);
+
+    // Get the result of the last operation.
+    fn get_operation_result(&self) -> isize;
+
+    fn clear_operation(&self);
+}
+
+// Get the static Flash object.
+pub fn get() -> &'static dyn Flash {
+    get_impl()
+}
+
+const DRIVER_NUMBER: usize = 0x40040;
```

Comment:
Why does this magic number have this value?  Should there be a reference (comment) here?

Reply:
I think this is just a bit of "tribal" Tock knowledge on where to look for this number. One would usually look for the board-specific implementation of Tock's `kernel::platform::Platform`, which for us is in `tock-on-titan/kernel/papa/src/main.rs`. This lists all the drivers supported for the particular board. From there you can find the magic numbers.